### ISSUE TITLE
FOCS Parser becomes very slow for nested parentheses

### ISSUE THREAD

**Issue by Morlic-fo** [freeorion/freeorion#1824]
### Environment
* **FreeOrion Version**:  test build 2017-10-23.0ac4a07
* **Operating System**: Win 7 / 64 bit
* **Fetched as**
  * Weekly development build

### Description
The FOCS parser takes a *very* long time to parse files that have a high level of nested parentheses. In current content, this is most noticeable for the experimentor outpost.

### Steps to reproduce
* Select an arbitrary focs file (e.g. default/scripting/buildings/EXPERIMENTOR_OUTPOST.focs.txt)
* Add parentheses around existing parentheses to increase number of nesting. For example,
```
EXPERIMENTOR_SPAWN_START_TURN
'''(((((((((([[EXPERIMENTOR_SPAWN_CALC_A]] * (1.2 - ( GalaxyPlanetDensity / 10 )) * ((Max(1, GalaxySize / 200))^0.4)))))))))))'''
```
* Start a game
* Check logs and compare time to parse the file.
* If the game still started in an acceptable amount of time, add another pair of parentheses to increase the nesting level and compare the timing to parse the file.

Result of above code:
```
23:40:00.742046 [debug] server : Pending.h:53 : Waiting for parse of "buildings" to complete.
23:40:01.742103 [debug] server : Pending.h:53 : Waiting for parse of "buildings" to complete.
(... some hundred lines of this...)
23:48:50.012319 [debug] server : Pending.h:53 : Waiting for parse of "buildings" to complete.
23:48:51.022376 [debug] server : Pending.h:53 : Waiting for parse of "buildings" to complete.
```
and still going strong after 9 minutes

**Comment by geoffthemedio** [comment]
Can you find an earlier version where this was not a problem?

**Comment by LGM-Doyle** [comment]
@Morlic-fo thanks for reporting this.

@geoffthemedio I checked back to at least May 2017 and the same problem exists.

Part of the problem is that when the spirit parser backtracks it leaks memory.  Adding parentheses, increases the backtracking, more memory leaks and more time is consumed.  You can verify this by checking the amount of memory used by the server, client and AI processes as you increase the number of redundant parentheses.

I had a work in progress branch fixing the memory leaks in the parser, but this bug allowed me to see that my approach was wrong.  I've started over and it may take some time to fix the leaks completely.

I can't tell if the leaks are masking an additional problem with the parentheses nesting.

**Comment by Morlic-fo** [comment]
Fixed by #1844
